code/modules/bluespace_exploration/encounter/ship_datum.dm

```diff
@@ -0,0 +1,161 @@
+	// - That NPC trading ship is now declared hostile to the player ships faction.
+	// - The cycle continues.
+	//Note: Doesn't take into account subtypes.
+	var/rogue_factions
+
+/datum/ship_datum/New()
+	. = ..()
+	//Really this should share a faction rather than have a damned new faction for EVERY ship, but this makes it easier to declare hostilities, so meh not on my todo list.
+	ship_faction = new ship_faction
+	rogue_factions = list()
+
+/datum/ship_datum/proc/update_ship()
+	if(critical)
+		return
+	//If somehow the docking port manages to get destroyed, assume the ship to be lost
+	var/obj/docking_port/mobile/M = SSshuttle.getShuttle(mobile_port_id, FALSE)
+	if(!M)
+		log_runtime("Unable to locate docking port [mobile_port_id]")
+		qdel(src)
+		return
+	if(M.mode != SHUTTLE_IDLE)
+		return
+	//Calculate Health
+	var/list/turfs = M.return_turfs()
+	calculate_health(turfs)
+	//Boom
+	if(integrity_remaining < 0)
+		hull_critical(turfs)
+	return turfs
+
+/datum/ship_datum/proc/calculate_health(list/turfs)
+	var/calculated_health = 0
+	for(var/turf/T in turfs)
+		if(istype(T, /turf/closed/wall))
+			calculated_health += 5		//Walls are always 5 health
+		else if(istype(T, /turf/open/floor))
+			//Broken plating - 1, plating - 2, broken floor - 3, floor - 4
+			var/turf/open/floor/F = T
+			calculated_health += 1
+			if(!F.broken && !F.burnt)
+				calculated_health += 1
+			if(!istype(T, /turf/open/floor/plating))
+				calculated_health += 2
+	if(calculated_health > max_ship_integrity)
+		max_ship_integrity = calculated_health
+	current_ship_integrity = calculated_health
+	var/difference = max_ship_integrity - current_ship_integrity
+	var/used_health_pool = max_ship_integrity * health_percentage
+	integrity_remaining = used_health_pool - difference
+
+/datum/ship_datum/proc/hull_critical(list/turfs)
+	if(critical)
+		return
+	critical = TRUE
+	if(!LAZYLEN(turfs))
+		WARNING("Ship [ship_name], port_id [mobile_port_id] has no turfs.")
+		return
+	var/turf/first_turf = turfs[1]
+	message_admins("The [ship_name], port_id [mobile_port_id] has been destroyed, at [ADMIN_JMP(first_turf)]")
+	log_attack("The [ship_name], port_id [mobile_port_id] has been destroyed!")
+	var/obj/docking_port/mobile/M = SSshuttle.getShuttle(mobile_port_id, FALSE)
+	if(!M)
+		WARNING("Warning, no docking port located on shuttle [ship_name]")
+	for(var/turf/T as() in turfs)
+		var/obj/machinery/light/L = locate() in T
+		if(L)
+			L.force_emergency_mode = TRUE
+			L.update()
+	//Play an alarm to anyone / any observers on the shuttle
+	for(var/client/C as() in GLOB.clients)
+		var/mob/client_mob = C.mob
+		if(get_turf(client_mob in turfs))
+			SEND_SOUND(C, 'sound/machines/alarm.ogg')
+	addtimer(CALLBACK(src, .proc/destroy_ship, turfs), 140)
+
+/datum/ship_datum/proc/destroy_ship(list/turfs)
+	set waitfor = FALSE
+	GLOB.ships_destroyed ++
+	var/exploded = FALSE
+	for(var/turf/T in turfs)
+		var/obj/machinery/bluespace_drive/BS = locate() in T
```

Comment:
Don't ship datums store a ref to the drive?

I feel like storing refs to the drive and the APCs would be cheaper than this loop existing.

Reply:
They do not store a ref to the drive, this stuff is only ever searched once when the ship is destroyed, so adding it in init would just move the same list checking to a different point.